internal/manifest/list_test.go

```diff
@@ -137,3 +137,96 @@ func TestChooseInstance(t *testing.T) {
 		}
 	}
 }
+
+func TestChooseInstanceByCompressionAndListInternalFromBlob(t *testing.T) {
```

Comment:
All of this is just copied `TestChooseInstance` to ensure that `ChooseInstanceByCompression` behaves the same, and doesn’t depend on the compression preference, isn’t it?

If so, please don‘t duplicate such a large piece of code; just loop inside the existing `TestChooseInstance`:
```go
chooseInstanceCalls := []func(sys *types.SystemContext, rawManifest []byte) (digest.Digest, error) {
   func(sys *types.SystemContext, rawManifest []byte) (digest.Digest, error) {
      list, err := ListPublicFromBlob(rawManifest, GuessMIMEType(rawManifest))
      …
      return list.ChooseInstance(sys)
   },
   func /* use ChooseInstanceByCompression … OptionalBoolTrue */,
   func /* use ChooseInstanceByCompression … OptionalBoolFalse */,
   func /* use ChooseInstanceByCompression … OptionalBoolUndefined, adding that is now cheap */,
}
…
for _, manifestList := range []struct {…} {
		for _, match := range manifestList.matchedInstances {
		    for callIndex, chooseInstance := range chooseInstanceCalls {
		        testName := …
		        digest, err = chooseInstance(&systemContext{…}, rawManifest
```

Reply:
I’d still advocate for this loop structure — or even shorter
```go
…
		for callIndex, chooseInstance := range chooseInstanceCalls {
			for _, match := range manifestList.matchedInstances {
			for _, arch := range manifestList.unmatchedInstances {
```
so that there is only one `chooseInstanceCalls` loop.

In my experiment it is shorter (-16 lines), it has more coverage (`OptionalBoolUndefined`), and it is much clearer about how the repeated instances of the code do / don’t differ, making it less likely that future updates will only change one of the sites.

I guess non-blocking.

Reply:
Fixed.